### DISCUSSION TITLE
Support for SQL type BIT

### DISCUSSION THREAD

**Discussion by junbl** [SeaQL/sea-orm#1606] [Ideas]
Currently, the SQL type BIT is only partially supported--it's in `sea_query::ColumnType`, but it doesn't get generated properly by `sea-orm-cli` and it doesn't get serialized with `into_json`.

For a table with a field like this:
```sql
  `field` bit(1) NOT NULL,
```

`sea` just generates a "custom" field:
```rust
    #[sea_orm(column_type = "custom(\"BIT(1)\")")]
    pub field: String,
```

Retrieving data with this model fails with the following error:
```
Query(SqlxError(ColumnDecode { index: "\"field\"", source: "mismatched types; Rust type `core::option::Option<alloc::string::String>` (as SQL type `VARCHAR`) is not compatible with SQL type `BIT`" }))'
```

Changing the rust type to `bool` (if `BIT(1)`) or `u32` or `u64`  (for larger numbers of bits) works fine for the model, but silently fails to be converted to `serde_json::Value` using `into_json`-- the resulting `Object`s are simply missing the fields.

An example repo to show the problem: <https://github.com/e-rhodes/sea-orm-test/tree/bit>

What should be done in this case?

Ideally, in my opinion, it would choose between `bool` or `u{x}` for the Rust type as appropriate for the number of bits (so `BIT(1)` -> `bool`, `BIT(2-8)` => `u8`, `BIT(9-16)` -> `u16`, etc.). Even just `bool` or `u64` would be helpful, since `BIT(1)` is likely the most common. Automatically generating `u64` but allowing retrieval to a `bool` after changing it manually would also be acceptable (similar to the current behavior with `TINYINT`).

The maximum value for the number of bits is 64 in MySQL ([source](https://dev.mysql.com/doc/refman/8.0/en/bit-type.html)), which means it could always go into a u64, but it doesn't look like there's a limit in postgres, which complicates things ([source](https://www.postgresql.org/docs/current/datatype-bit.html)).

If there isn't a limit for Postgres, then maybe it would have to be `Vec<bool>`. For `BIT` as opposed to `BIT VARYING`, it would be really cool to generate a fixed array (i.e. `BIT(m)` -> `[bool; m]`). That way you can just destructure the array without having to check the length in the Rust code when you know it's guaranteed by the database. This might actually be a nicer interface to use than `u64`. 

It doesn't seem like the `column_type` attribute really has an effect. Using the correct `sea_query::ColumnType::Bit` doesn't change the output, only changing the Rust type fixes the error:

```rust
    #[sea_orm(column_type = "Bit(Some(1))")]
    pub field: bool,
```

So it seems like it should be left off entirely, but maybe it's helpful? I'm not familiar enough with the inner workings to know.

As for `into_json`, the current behavior of silently failing, even when the model would deserialize correctly, is definitely not ideal lol. I think at the very least it should produce an error when it fails to match a type (just simply adding a `return Err()` at then end of all the `try_get_type!`s in `impl FromQueryResult for JsonValue`). Ideally, it would go to `Value::Bool` or `Value::Number` as appropriate (or `Value::Array(Vec<Value::Bool>`).

**Comment by billy1624** [comment]
Hey @e-rhodes, thanks for the proposal!! I think using [bit-vec](https://crates.io/crates/bit-vec) to represent `BIT` datatype is more appropriate. However, in SQLx, the encoding and decoding of it is only supported by Postgres backend.

**Reply by billy1624** [reply]
I do have a temporary workaround for you :)
- https://github.com/SeaQL/sea-orm/pull/1612

**Reply by junbl** [reply]
Nice, I wasn't aware of that crate--looks great! I did also see the similarly named [bitvec](https://docs.rs/bitvec/latest/bitvec/) which is >1.0 and seems to have more features (includes both vector and array types and more methods) but not a super friendly interface.

There's also [bit-array](https://docs.rs/bit-array/latest/bit_array/), which looks like the same API as bit-vec, but for constant size arrays. Doesn't seem to be a way in any to do the compile-checked destructuring into `N` `bool`s, which is a shame but understandable.


Thanks so much for that example!! I wasn't able to get it working with MySQL, unfortunately. I changed `select_as` to `"UNSIGNED"` from `BIGINT`, to match MySQL's CAST syntax, but this creates a field with the name `CAST(`table`.`field` AS UNSIGNED)` in the resulting data, which obviously won't match the expected name `field` for the model deserialization. Is there a way to add an alias as an attribute macro, or should SeaORM handle that automatically whenever the select_as option is used? Regardless, since `.into_json` bypasses anything set on the `Model`, the `select_as` workaround won't affect it.
